src/Unsafe/Linear.hs

```diff
@@ -20,42 +26,138 @@
 -- it.
 
 module Unsafe.Linear
-  ( -- * Unsafe Coersions
+  ( -- * Unsafe Coercions
     coerce,
     toLinear,
     toLinear2,
     toLinear3,
+    toLinearN,
+    ToLinearN (..),
   )
   where
 
-import qualified Unsafe.Coerce as NonLinear
-import GHC.Exts (TYPE, RuntimeRep)
+import Data.Type.Equality (type (~~))
+import Data.Kind (Constraint)
+import Unsafe.Coerce (UnsafeEquality (..), unsafeEqualityProof)
+import GHC.Exts (TYPE, RuntimeRep (..))
+import GHC.TypeNats
 
 -- | Linearly typed @unsafeCoerce@
 coerce :: forall a b. a %1-> b
 coerce a =
-  case NonLinear.unsafeEqualityProof @a @b of
-    NonLinear.UnsafeRefl -> a
+  case unsafeEqualityProof @a @b of
+    UnsafeRefl -> a
 {-# INLINE coerce #-}
 
+-- TODO: In GHC 9.0, using toLinearN to implement toLinear2 and toLinear3
+-- produces more Core terms, though fewer types and coercions. Does this hamper
+-- inlining? I imagine it might. I suspect this is still the case for 9.2, but
+-- I haven't checked. If we want, we might be able to reduce
+-- unsafeLinearityProofN to one case expression by using some (quantified
+-- constraint?) tricks.
```

Comment:
This would really be nice but I don't see how to do this at the moment. And I'm not sure this deserves to spend more time on.

Reply:
Yeah, not planning to touch it any more for now. Sliding back toward the `linear-generics` thing.